File: alpha_parser/alpha_lexer.py
from typing import List, Optional
import re
from .tokens import Token, TokenType, KNOWN_FUNCTIONS, KNOWN_VARIABLES

class AlphaLexer:
    def __init__(self, text: str):
        self.text = text
        self.pos = 0
        self.current_char = self.text[0] if self.text else None
        
    def advance(self):
        self.pos += 1
        if self.pos < len(self.text):
            self.current_char = self.text[self.pos]
        else:
            self.current_char = None
            
    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.advance()
            
    def number(self):
        result = ''
        while self.current_char is not None and (self.current_char.isdigit() or self.current_char == '.'):
            result += self.current_char
            self.advance()
        return float(result)
    
    def identifier(self):
        result = ''
        while self.current_char is not None and (self.current_char.isalnum() or self.current_char == '_'):
            result += self.current_char
            self.advance()
        return result
    
    def get_next_token(self) -> Token:
        while self.current_char is not None:
            if self.current_char.isspace():
                self.skip_whitespace()
                continue
                
            if self.current_char.isdigit():
                return Token(TokenType.NUMBER, self.number())
                
            if self.current_char.isalpha() or self.current_char == '_':
                ident = self.identifier()
                if ident in ['ts_min', 'ts_max', 'ts_rank', 'ts_sum', 'ts_mean', 'ts_std', 'ts_corr', 'ts_cov']:
                    return Token(TokenType.FUNCTION, ident)
                return Token(TokenType.IDENTIFIER, ident)
                
            if self.current_char == '+':
                self.advance()
                return Token(TokenType.PLUS)
                
            if self.current_char == '-':
                self.advance()
                return Token(TokenType.MINUS)
                
            if self.current_char == '*':
                self.advance()
                if self.current_char == '*':
                    self.advance()
                    return Token(TokenType.POWER)
                return Token(TokenType.MULTIPLY)
                
            if self.current_char == '/':
                self.advance()
                return Token(TokenType.DIVIDE)
                
            if self.current_char == '=':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    return Token(TokenType.EQUAL)
                return Token(TokenType.EQUAL)
                
            if self.current_char == '!':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    return Token(TokenType.NOT_EQUAL)
                return Token(TokenType.NOT)
                
            if self.current_char == '>':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    return Token(TokenType.GREATER_EQUAL)
                return Token(TokenType.GREATER)
                
            if self.current_char == '<':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    return Token(TokenType.LESS_EQUAL)
                return Token(TokenType.LESS)
                
            if self.current_char == '&':
                self.advance()
                if self.current_char == '&':
                    self.advance()
                    return Token(TokenType.AND)
                
            if self.current_char == '|':
                self.advance()
                if self.current_char == '|':
                    self.advance()
                    return Token(TokenType.OR)
                
            if self.current_char == '(':
                self.advance()
                return Token(TokenType.LPAREN)
                
            if self.current_char == ')':
                self.advance()
                return Token(TokenType.RPAREN)
                
            if self.current_char == '[':
                self.advance()
                return Token(TokenType.LBRACKET)
                
            if self.current_char == ']':
                self.advance()
                return Token(TokenType.RBRACKET)
                
            if self.current_char == ',':
                self.advance()
                return Token(TokenType.COMMA)
                
            if self.current_char == '?':
                self.advance()
                return Token(TokenType.QUESTION)
                
            if self.current_char == ':':
                self.advance()
                return Token(TokenType.COLON)
                
            raise ValueError(f"Invalid character: {self.current_char}")
            
        return Token(TokenType.EOF)
    
    def get_all_tokens(self) -> List[Token]:
        """모든 토큰을 반환"""
        tokens = []
        while True:
            token = self.get_next_token()
            tokens.append(token)
            if token.type == TokenType.EOF:
                break
        return tokens 

File: alpha_parser/src/alpha_lexer.cpp
#include "alpha_lexer.hpp"
#include <stdexcept>
#include <cctype>
#include <algorithm>

namespace alpha_parser {

AlphaLexer::AlphaLexer(const std::string& text) 
    : text(text), pos(0), token_pos(0) {
    current_char = text.empty() ? '\0' : text[0];
    tokens = tokenize();
    token_pos = 0;
}

void AlphaLexer::advance() {
    pos++;
    if (pos >= text.length()) {
        current_char = '\0';
    } else {
        current_char = text[pos];
    }
}

char AlphaLexer::peek_char(int offset) const {
    size_t peek_pos = pos + offset;
    if (peek_pos >= text.length()) {
        return '\0';
    }
    return text[peek_pos];
}

Token AlphaLexer::number() {
    std::string result;
    
    // Handle numbers starting with decimal point
    if (current_char == '.') {
        result = "0";
        result += current_char;
        advance();
        while (current_char != '\0' && std::isdigit(current_char)) {
            result += current_char;
            advance();
        }
        return Token(TokenType::NUMBER, std::stod(result));
    }
    
    // Handle regular numbers
    while (current_char != '\0' && std::isdigit(current_char)) {
        result += current_char;
        advance();
    }
    
    if (current_char == '.') {
        result += current_char;
        advance();
        while (current_char != '\0' && std::isdigit(current_char)) {
            result += current_char;
            advance();
        }
    }
    
    return Token(TokenType::NUMBER, std::stod(result));
}

Token AlphaLexer::identifier() {
    std::string result;
    
    while (current_char != '\0' && (std::isalnum(current_char) || current_char == '_')) {
        result += current_char;
        advance();
    }
    
    return Token(TokenType::IDENTIFIER, result);
}

std::vector<Token> AlphaLexer::tokenize() {
    std::vector<Token> tokens;
    
    while (current_char != '\0') {
        if (std::isspace(current_char)) {
            advance();
        } else if (std::isalpha(current_char)) {
            // Check for IndClass pattern
            std::regex indclass_pattern("IndClass\\.(sector|industry|subindustry)");
            std::smatch match;
            std::string remaining = text.substr(pos);
            
            if (std::regex_search(remaining, match, indclass_pattern) && match.position() == 0) {
                std::string value = match.str();
                tokens.emplace_back(TokenType::IDENTIFIER, value);
                pos += value.length();
                current_char = (pos < text.length()) ? text[pos] : '\0';
                continue;
            } else {
                tokens.push_back(identifier());
            }
        } else if (std::isdigit(current_char) || (current_char == '.' && std::isdigit(peek_char()))) {
            tokens.push_back(number());
        } else if (current_char == '+') {
            tokens.emplace_back(TokenType::PLUS);
            advance();
        } else if (current_char == '-') {
            tokens.emplace_back(TokenType::MINUS);
            advance();
        } else if (current_char == '*') {
            if (peek_char() == '*') {
                advance();
                tokens.emplace_back(TokenType::POWER);
            } else {
                tokens.emplace_back(TokenType::MULTIPLY);
            }
            advance();
        } else if (current_char == '/') {
            tokens.emplace_back(TokenType::DIVIDE);
            advance();
        } else if (current_char == '^') {
            tokens.emplace_back(TokenType::POWER);
            advance();
        } else if (current_char == '%') {
            tokens.emplace_back(TokenType::MODULO);
            advance();
        } else if (current_char == '(') {
            tokens.emplace_back(TokenType::LPAREN);
            advance();
        } else if (current_char == ')') {
            tokens.emplace_back(TokenType::RPAREN);
            advance();
        } else if (current_char == ',') {
            tokens.emplace_back(TokenType::COMMA);
            advance();
        } else if (current_char == '[') {
            tokens.emplace_back(TokenType::LBRACKET);
            advance();
        } else if (current_char == ']') {
            tokens.emplace_back(TokenType::RBRACKET);
            advance();
        } else if (current_char == '<') {
            if (peek_char() == '=') {
                advance();
                tokens.emplace_back(TokenType::LESS_EQUAL);
            } else {
                tokens.emplace_back(TokenType::LESS);
            }
            advance();
        } else if (current_char == '>') {
            if (peek_char() == '=') {
                advance();
                tokens.emplace_back(TokenType::GREATER_EQUAL);
            } else {
                tokens.emplace_back(TokenType::GREATER);
            }
            advance();
        } else if (current_char == '=') {
            if (peek_char() == '=') {
                advance();
                tokens.emplace_back(TokenType::EQUAL);
            } else {
                throw std::runtime_error("Invalid character: " + std::string(1, current_char));
            }
            advance();
        } else if (current_char == '!') {
            if (peek_char() == '=') {
                advance();
                tokens.emplace_back(TokenType::NOT_EQUAL);
            } else {
                tokens.emplace_back(TokenType::NOT);
            }
            advance();
        } else if (current_char == '&') {
            if (peek_char() == '&') {
                advance();
                tokens.emplace_back(TokenType::AND);
                advance();
            } else {
                throw std::runtime_error("Invalid character: " + std::string(1, current_char) + " (did you mean && ?)");
            }
        } else if (current_char == '|') {
            if (peek_char() == '|') {
                advance();
                tokens.emplace_back(TokenType::OR);
                advance();
            } else {
                throw std::runtime_error("Invalid character: " + std::string(1, current_char) + " (did you mean || ?)");
            }
        } else if (current_char == '?') {
            tokens.emplace_back(TokenType::QUESTION);
            advance();
        } else if (current_char == ':') {
            tokens.emplace_back(TokenType::COLON);
            advance();
        } else {
            throw std::runtime_error("Invalid character: " + std::string(1, current_char));
        }
    }
    
    tokens.emplace_back(TokenType::EOF_TOKEN);
    return tokens;
}

Token AlphaLexer::get_next_token() {
    if (token_pos >= tokens.size()) {
        return Token(TokenType::EOF_TOKEN);
    }
    
    Token token = tokens[token_pos];
    token_pos++;
    return token;
}

Token AlphaLexer::peek(int k) const {
    size_t target = token_pos + k;
    if (target < tokens.size()) {
        return tokens[target];
    }
    return Token(TokenType::EOF_TOKEN);
}

std::vector<Token> AlphaLexer::get_all_tokens() const {
    return tokens;
}

void AlphaLexer::reset() {
    token_pos = 0;
}

} // namespace alpha_parser 